# ğŸ¢ Warehouse Rule

The Warehouse Rule validates data product warehouse configurations to ensure cost control and configuration integrity.

> **ğŸ¯ Quick Summary**: Auto-approves warehouse size reductions and valid configurations. Requires manual review for size increases and configuration errors.

## ğŸ“ When Does This Rule Apply?

The warehouse rule triggers when your MR modifies files matching these patterns:
- `**/product.yaml` (containing warehouse configurations)
- `**/product.yml`

**Examples of triggering files**:
- `dataproducts/analytics/helloaggregate/dev/product.yaml`
- `dataproducts/reporting/prod/product.yml`

### ğŸ§© YAML Section-Aware Validation

The warehouse rule now uses **line-level validation** and covers the entire warehouse file:

- **Entire file** - Complete warehouse configuration validation

**Example of what the rule validates**:
```yaml
---
name: helloaggregate           # âœ… Validated by warehouse rule
kind: aggregated               # âœ… Validated by warehouse rule
warehouses:                    # âœ… Validated by warehouse rule
- type: user                   #     (All lines in file)
  size: XSMALL                 
- type: service_account        
  size: XSMALL                 
service_account:               # âœ… Validated by warehouse rule  
  dbt: true                    #     (All lines in file)
tags:                          # âœ… Validated by warehouse rule
  data_product: helloaggregate # âœ… (All lines in file)
```

## âœ… Auto-Approved Changes

Your MR will be **automatically approved** if:

### âœ… **Warehouse Size Reductions**
Cost-saving changes are auto-approved:
- `LARGE` â†’ `MEDIUM`, `SMALL`, or `XSMALL`
- `MEDIUM` â†’ `SMALL` or `XSMALL`
- `SMALL` â†’ `XSMALL`

```yaml
# âœ… Auto-approved: Cost reduction in warehouses section
warehouses:
- type: user
  size: XSMALL    # Changed from SMALL
- type: service_account  
  size: SMALL     # Changed from MEDIUM
```

### âœ… **Valid Configuration Updates**
Well-formed configurations pass validation:
- Proper YAML syntax in validated sections
- Valid warehouse size values (`XSMALL`, `SMALL`, `MEDIUM`, `LARGE`)
- Correct warehouse section structure
- Service account configurations without issues

```yaml
# âœ… Auto-approved: Valid warehouse section configuration
name: "analytics-data-product"    # â­ï¸ Not warehouse rule responsibility
warehouses:                       # âœ… Warehouse rule validates this
- type: user
  size: MEDIUM                    # Valid size value
- type: service_account
  size: SMALL                     # Valid size value
service_account:                  # âœ… Warehouse rule validates this
  dbt: true                       # Valid service account config
```

### âœ… **Complete File Validation**
With line-level validation, the warehouse rule validates the entire warehouse file:
- All changes in `product.yaml` files are validated by the warehouse rule
- The rule provides complete coverage for warehouse configuration files
- All sections affect this rule's decision

## âš ï¸ Manual Review Required

Your MR will require **manual review** if:

### âš ï¸ **Warehouse Size Increases**
Cost-impacting changes need approval:
- `XSMALL` â†’ `SMALL`, `MEDIUM`, or `LARGE`
- `SMALL` â†’ `MEDIUM` or `LARGE`
- `MEDIUM` â†’ `LARGE`

```yaml
# âš ï¸ Manual review: Cost increase in warehouse section
warehouses:
- type: user
  size: LARGE     # Changed from MEDIUM
- type: service_account
  size: MEDIUM    # Changed from SMALL
```

**Why?** Size increases have budget implications and require cost approval.

### âš ï¸ **Configuration Issues**

#### Malformed YAML Syntax
```yaml
# âŒ Invalid: Missing quotes for special characters
name: Data Product: Analytics

# âœ… Valid: Properly quoted
name: "Data Product: Analytics"
```

#### Invalid Warehouse Size Values
```yaml
# âŒ Invalid: Lowercase not supported
warehouse: medium

# âœ… Valid: Uppercase required
warehouse: MEDIUM
```

#### Missing Required Fields
```yaml
# âŒ Invalid: Missing required fields
warehouse: LARGE

# âœ… Valid: Complete configuration
name: "my-data-product"
warehouse: LARGE
description: "Product description"
owner: "team@company.com"
```

### âš ï¸ **File Access Problems**
Technical issues trigger manual review:
- Cannot fetch file content from GitLab
- Network timeouts or API errors
- File permission issues

### Configuration Validation Flow

```mermaid
flowchart TD
    A[ğŸ“‹ Warehouse Configuration] --> B{ğŸ” Required Fields Check}
    B -->|âŒ Missing| C[âš ï¸ Missing Fields Error<br/>â€¢ name<br/>â€¢ warehouse<br/>â€¢ description<br/>â€¢ owner]
    B -->|âœ… Complete| D{ğŸ“ Size Value Check}
    
    D -->|âŒ Invalid| E[âš ï¸ Invalid Size Error<br/>Must be: SMALL, MEDIUM, LARGE]
    D -->|âœ… Valid| F{ğŸ“ YAML Syntax Check}
    
    F -->|âŒ Invalid| G[âš ï¸ Syntax Error<br/>â€¢ Quotes missing<br/>â€¢ Indentation wrong<br/>â€¢ Special characters]
    F -->|âœ… Valid| H[âœ… Configuration Valid]
    
    classDef errorNode fill:#f8d7da,stroke:#721c24,stroke-width:2px,color:#721c24
    classDef successNode fill:#d4edda,stroke:#155724,stroke-width:2px,color:#155724
    classDef processNode fill:#e2e3e5,stroke:#6c757d,stroke-width:2px,color:#495057
    
    class C,E,G errorNode
    class H successNode
    class A,B,D,F processNode
```

### File Processing Pipeline

```mermaid
sequenceDiagram
    participant U as ğŸ‘¤ User
    participant G as ğŸ“ GitLab
    participant N as ğŸ¤– Naysayer
    participant R as ğŸ¢ Warehouse Rule
    
    U->>G: ğŸ“¤ Push warehouse changes
    G->>N: ğŸ”” Webhook: MR created/updated
    N->>R: ğŸ¯ Check if rule applies
    R->>R: ğŸ” Analyze file patterns
    
    alt ğŸ“ Warehouse file detected
        R->>G: ğŸ“¥ Fetch file content
        G-->>R: ğŸ“„ Return file content
        R->>R: ğŸ” Parse YAML syntax
        R->>R: âœ… Validate configuration
        R->>R: ğŸ“Š Analyze size changes
        
        alt ğŸ“‰ Size reduction or no change
            R->>N: âœ… Auto-approve
            N->>G: ğŸ‰ Approve MR
            G->>U: âœ… MR approved automatically
        else ğŸ“ˆ Size increase
            R->>N: âš ï¸ Manual review required
            N->>G: ğŸ” Request manual review
            G->>U: âš ï¸ Manual review needed
        end
    else ğŸ“ No warehouse files
        R->>N: â­ï¸ Skip rule
        N->>G: â¡ï¸ Continue with other rules
    end
```

## ğŸ”§ Troubleshooting

### Common Error Messages

#### "Invalid warehouse size"
**Cause**: Using unsupported warehouse size value  
**Solution**: Use only `SMALL`, `MEDIUM`, or `LARGE` (uppercase)

```yaml
# âŒ These are invalid
warehouse: small
warehouse: Medium  
warehouse: XL

# âœ… These are valid
warehouse: SMALL
warehouse: MEDIUM
warehouse: LARGE
```

#### "Malformed YAML"
**Cause**: YAML syntax errors  
**Solution**: Validate YAML syntax before committing

```yaml
# âŒ Invalid: Unquoted special characters
description: Cost: $500/month

# âœ… Valid: Properly quoted
description: "Cost: $500/month"
```

#### "Missing required fields"
**Cause**: Required configuration fields not present  
**Solution**: Include all mandatory fields

```yaml
# âœ… Minimum required configuration
name: "product-name"           # Required
warehouse: MEDIUM              # Required  
description: "Brief description" # Required
owner: "team@company.com"      # Required
```

#### "Failed to fetch file"
**Cause**: File access or network issues  
**Solutions**:
1. Check file exists at the correct path
2. Verify GitLab permissions
3. Retry if temporary network issue
4. Contact platform team if persistent

### Validation Steps

1. **Check file path**: Ensure file is in `dataproducts/*/product.yaml` format
2. **Validate YAML**: Use online YAML validator or `yamllint`
3. **Verify size value**: Must be exactly `XSMALL`, `SMALL`, `MEDIUM`, `LARGE`, etc.
4. **Include required fields**: name, warehouses section with proper configuration
5. **Test locally**: Parse YAML to catch syntax issues early

## âš™ï¸ Configuration

### Environment Variables

```bash
# Enable/disable warehouse validation
WAREHOUSE_RULE_ENABLED=true

# Allow warehouse size increases (bypasses cost approval)
WAREHOUSE_ALLOW_SIZE_INCREASES=false

# Maximum file size to process (bytes)
WAREHOUSE_MAX_FILE_SIZE=1048576  # 1MB

# Strict mode (additional validations)
WAREHOUSE_STRICT_MODE=false

# Debug logging
WAREHOUSE_DEBUG=false
```

### Default Configuration

```yaml
# Default settings applied if not specified
warehouse_rule:
  enabled: true
  allow_size_increases: false
  max_file_size: 1048576
  strict_mode: false
  required_fields:
    - name
    - warehouse  
    - description
    - owner
  valid_sizes:
    - SMALL
    - MEDIUM
    - LARGE
```

## ğŸ“Š Rule Behavior

### Decision Logic Flow

```mermaid
graph TB
    A[ğŸ“ MR File Changes] --> B{ğŸ¢ Warehouse YAML File?}
    B -->|âŒ No| C[â­ï¸ Skip Rule<br/>Not applicable]
    B -->|âœ… Yes| D[ğŸ” Fetch File Content]
    
    D --> E{ğŸ“¥ File Access OK?}
    E -->|âŒ No| F[âš ï¸ Manual Review<br/>ğŸš« Access Error]
    E -->|âœ… Yes| G{ğŸ“ Valid YAML Syntax?}
    
    G -->|âŒ No| H[âš ï¸ Manual Review<br/>ğŸš« Syntax Error]
    G -->|âœ… Yes| I[ğŸ” Analyze Configuration]
    
    I --> J{ğŸ“Š Warehouse Size Change?}
    J -->|ğŸ”„ No Change| K{âœ… Valid Config?}
    J -->|ğŸ“‰ Size Reduction| L[ğŸ’° Cost Savings Detected]
    J -->|ğŸ“ˆ Size Increase| M[âš ï¸ Manual Review<br/>ğŸ’¸ Cost Impact Review]
    
    L --> K
    K -->|âŒ Invalid| N[âš ï¸ Manual Review<br/>ğŸš« Config Issues]
    K -->|âœ… Valid| O[âœ… Auto-Approve<br/>ğŸ‰ Changes Approved]
    
    classDef approveNode fill:#d4edda,stroke:#155724,stroke-width:2px,color:#155724
    classDef reviewNode fill:#f8d7da,stroke:#721c24,stroke-width:2px,color:#721c24
    classDef processNode fill:#e2e3e5,stroke:#6c757d,stroke-width:2px,color:#495057
    classDef skipNode fill:#fff3cd,stroke:#856404,stroke-width:2px,color:#856404
    
    class O approveNode
    class F,H,M,N reviewNode
    class A,D,I,L processNode
    class C skipNode
```

### Cost Impact Analysis

```mermaid
graph LR
    subgraph "ğŸ’° Cost Impact Matrix"
        A[SMALL] -.->|ğŸ“ˆ +Cost| B[MEDIUM]
        A -.->|ğŸ“ˆ ++Cost| C[LARGE]
        B -->|ğŸ“‰ -Cost| A
        B -.->|ğŸ“ˆ +Cost| C
        C -->|ğŸ“‰ -Cost| A
        C -->|ğŸ“‰ -Cost| B
    end
    
    subgraph "ğŸ¯ Decision Rules"
        D[ğŸ“‰ Size Reduction<br/>Auto-Approve] 
        E[ğŸ“ˆ Size Increase<br/>Manual Review]
    end
    
    A -.->|Increase| E
    B -.->|Increase| E
    B -->|Decrease| D
    C -->|Decrease| D
    
    classDef increase fill:#f8d7da,stroke:#721c24,stroke-width:2px,color:#721c24
    classDef decrease fill:#d4edda,stroke:#155724,stroke-width:2px,color:#155724
    classDef neutral fill:#e2e3e5,stroke:#6c757d,stroke-width:2px
    
    class E increase
    class D decrease
    class A,B,C neutral
```

### Cost Impact Matrix

| **From** | **To** | **Cost Impact** | **Decision** | **Reason** |
|----------|--------|----------------|--------------|------------|
| ğŸ”¹ SMALL | ğŸ”¸ MEDIUM | ğŸ“ˆ +Cost | âš ï¸ Manual Review | Cost increase requires approval |
| ğŸ”¹ SMALL | ğŸ”¶ LARGE | ğŸ“ˆ ++Cost | âš ï¸ Manual Review | Significant cost increase |
| ğŸ”¸ MEDIUM | ğŸ”¹ SMALL | ğŸ“‰ -Cost | âœ… Auto-Approve | Cost reduction approved |
| ğŸ”¸ MEDIUM | ğŸ”¶ LARGE | ğŸ“ˆ +Cost | âš ï¸ Manual Review | Cost increase requires approval |
| ğŸ”¶ LARGE | ğŸ”¹ SMALL | ğŸ“‰ --Cost | âœ… Auto-Approve | Significant cost reduction |
| ğŸ”¶ LARGE | ğŸ”¸ MEDIUM | ğŸ“‰ -Cost | âœ… Auto-Approve | Cost reduction approved |

### Configuration Examples Comparison

```mermaid
graph TB
    subgraph "âœ… Valid Configurations"
        A["ğŸ“‹ Complete Config<br/>---<br/>name: 'analytics-pipeline'<br/>warehouse: MEDIUM<br/>description: 'Analytics processing'<br/>owner: 'team@company.com'"]
        B["ğŸ“‹ Minimal Valid<br/>---<br/>name: 'simple-service'<br/>warehouse: SMALL<br/>description: 'Basic service'<br/>owner: 'dev@company.com'"]
    end
    
    subgraph "âŒ Invalid Configurations"
        C["ğŸš« Missing Fields<br/>---<br/>warehouse: LARGE<br/># Missing: name, description, owner"]
        D["ğŸš« Invalid Size<br/>---<br/>name: 'test'<br/>warehouse: medium<br/># Should be: MEDIUM"]
        E["ğŸš« YAML Syntax Error<br/>---<br/>name: Data: Analytics<br/># Missing quotes for special chars"]
    end
    
    classDef validConfig fill:#e8f5e8,stroke:#388e3c,stroke-width:2px,color:#155724
    classDef invalidConfig fill:#ffebee,stroke:#d32f2f,stroke-width:2px,color:#721c24
    
    class A,B validConfig
    class C,D,E invalidConfig
```

## ğŸ¯ Best Practices

### Writing Warehouse Configurations

```yaml
# âœ… Good example
name: "analytics-pipeline"
warehouse: MEDIUM
description: "Daily analytics data processing pipeline"
owner: "analytics-team@company.com"
environment: "production"
cost_center: "data-analytics"
schedule: "0 2 * * *"  # Daily at 2 AM
```

### Size Selection Guidelines

- **SMALL**: Development, testing, small datasets
- **MEDIUM**: Production workloads, moderate datasets  
- **LARGE**: High-volume processing, large datasets

### Change Management

1. **Size increases**: Prepare business justification before requesting
2. **Documentation**: Update descriptions when changing configurations
3. **Testing**: Validate changes in development environment first
4. **Monitoring**: Track usage and costs after size changes

## ğŸ†˜ Getting Help

### When to Contact Support

- Persistent validation failures after fixing syntax
- Questions about appropriate warehouse sizing
- Issues with file access or permissions
- Need emergency size increase approval

### Information to Include

- **MR URL**: Link to blocked merge request
- **File path**: Exact path to product.yaml file
- **Error message**: Complete error text from rule
- **Configuration**: Current and desired product.yaml content
- **Business justification**: For size increase requests

### Emergency Procedures

For urgent production issues requiring immediate size increases:

1. Contact on-call team with justification
2. Request temporary bypass if available
3. Follow up with proper approval process
4. Document incident for review

## ğŸ“ˆ Monitoring

### Key Metrics

- **Auto-approval rate**: Percentage of changes approved automatically
- **Size increase requests**: Frequency and justification quality
- **Configuration errors**: Common syntax and validation issues
- **Cost impact**: Total cost changes from approved size increases

### Performance Targets

- **Rule execution time**: < 5 seconds per file
- **False positive rate**: < 2% of valid configurations blocked
- **Auto-approval rate**: > 85% of warehouse changes

## ğŸ“š Related Documentation

- [Rule Creation Guide](../RULE_CREATION_GUIDE.md) - For developers
- [Configuration Management](../CONFIG_MANAGEMENT.md) - Global settings
- [Cost Management](../COST_MANAGEMENT.md) - Warehouse sizing guidelines

---

**ğŸ’¡ Pro Tip**: Most warehouse rule issues can be resolved by ensuring proper YAML syntax and using exact uppercase values for warehouse sizes (`SMALL`, `MEDIUM`, `LARGE`).